### PR TITLE
Job Features & Other Fixes

### PR DESCRIPTION
**Pull request by alexandrustoica**
## Important Read All 

- [x] I fixed the SkillController (Hibernate was unable to return me the list of skills from provider)
I had to do it using a query with Spring Data
- [x] I fixed the RequestController -- provider.equals(provider) 
- [x] I added a new way to apply to a job (by using the job's id and a simple message)
- [x] I added a new way to accept a provider to a job (without having to "update" the request object from the database) .. all you have to do now is to give the API the job's ID and the provider's ID
- [x] I added almost ~ all the ways you can search for a job on the website (by location, by category, by expiration_date before x, by type, etc)

### CODE REVIEW

**Comment by alexandrustoica** [comment]
## Important

**Before we added SpringSecurity I tested my SkillController (insert/delete for skill) and it worked fine with provider.getSkills().add(skill) and providerRepo.save(provider) <-- this no longer works, Hibernate is unable to load the lazy set from provider ... IDK why.**

File: src/main/java/meeseeks/box/controller/JobController.java [hunk 1]
```diff
@@ -51,10 +37,10 @@
 
     @Autowired
     public JobController(final JobRepository jobRepository,
-            final ProviderRepository providerRepository,
-            final ConsumerRepository consumerRepository,
-            final CategoryRepository categoryRepository,
-            final AvailabilityRepository availabilityRepository) {
+                         final ProviderRepository providerRepository,
```
Comment (decebaldecebal) [inline src/main/java/meeseeks/box/controller/JobController.java]: Idk why you use so many indents, what was before is ok.Take a look at the Java coding standards:
http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-136091.html#248

File: src/main/java/meeseeks/box/controller/RequestController.java [hunk 2]
```diff
@@ -39,9 +40,8 @@ public RequestController(final RequestRepository requestRepository,
     @Secured({"ROLE_PROVIDER"})
     @PostMapping("/insert")
     public ResponseEntity<RequestEntity> insert(@Valid @RequestBody RequestEntity request,
```
Comment (decebaldecebal) [inline src/main/java/meeseeks/box/controller/RequestController.java]: I believe we can now delete the insert action since we have applyToJob and now represents a security risk since multiple requests can be added using this action for the same job/provider combo.

File: src/main/java/meeseeks/box/controller/RequestController.java [hunk 3]
```diff
@@ -64,28 +63,30 @@ public RequestController(final RequestRepository requestRepository,
     @Secured({"ROLE_PROVIDER"})
     @PatchMapping("/update")
     public ResponseEntity<RequestEntity> update(@Valid @RequestBody RequestEntity updated,
-                                                final Authentication authentication) {
-        ProviderEntity user = (ProviderEntity) authentication.getPrincipal();
-        return user.equals(updated.getProvider()) ?
+                                                @AuthenticationPrincipal @ApiIgnore ProviderEntity provider) {
+        return provider.getId().equals(updated.getProvider().getId()) ?
                 new ResponseEntity<>(requestRepository.save(updated), HttpStatus.OK) :
```
Comment (decebaldecebal) [inline src/main/java/meeseeks/box/controller/RequestController.java]: You should first get the old request from the database by using an id which should be in the patch, and then update the request since this was a new request is created since the request which comes from the body is not managed by hibernate.

File: src/main/java/meeseeks/box/domain/RequestEntity.java [hunk 4]
```diff
@@ -87,7 +94,7 @@ public void setMessage(final @NotNull String message) {
         this.message = message;
     }
 
-    public Boolean isAccepted() {
+    public Boolean getAccepted() {
```
Comment (decebaldecebal) [inline src/main/java/meeseeks/box/domain/RequestEntity.java]: It is recommended for boolean getters to be named with "is" instead of "get".

File: src/test/java/meeseeks/box/controller/SecurityTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package meeseeks.box.controller;
+
+import meeseeks.box.domain.UserEntity;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.authority.AuthorityUtils;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+
+class SecurityTest {
```
Comment (decebaldecebal) [inline src/test/java/meeseeks/box/controller/SecurityTest.java]: Does this class have any purpose?

File: src/test/java/meeseeks/box/controller/SkillControllerTest.java [hunk 6]
```diff
@@ -1,112 +0,0 @@
...
-import static org.mockito.Matchers.notNull;
-import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
-import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
-
-@RunWith(SpringRunner.class)
-@WebMvcTest(SkillController.class)
-@ContextConfiguration
-public class SkillControllerTest {
```
Comment (decebaldecebal) [inline src/test/java/meeseeks/box/controller/SkillControllerTest.java]: Why was this deleted?

File: src/main/java/meeseeks/box/domain/RequestEntity.java (hunk 4 above)
Comment (alexandrustoica) [reply src/main/java/meeseeks/box/domain/RequestEntity.java]: We will use get because @Query needs getters and setters as a default way of getting / setting fields in your proxy entity.

File: src/test/java/meeseeks/box/controller/SecurityTest.java (hunk 5 above)
Comment (alexandrustoica) [reply src/test/java/meeseeks/box/controller/SecurityTest.java]: Yes, will help me to write better tests for Controllers. -- it's still a WIP on that side

File: src/test/java/meeseeks/box/controller/SkillControllerTest.java (hunk 6 above)
Comment (alexandrustoica) [reply src/test/java/meeseeks/box/controller/SkillControllerTest.java]: Because since we use @AuthPrincipal I have to find a way to test it here. This test no longer works, I have to find an alternative way of testing this. This test is also useless because I plan on moving the logic in services in order to test it faster (--now that we have Swagger)

File: src/main/java/meeseeks/box/controller/JobController.java (hunk 1 above)
Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/JobController.java]: vim

File: src/main/java/meeseeks/box/controller/RequestController.java (hunk 3 above)
Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/RequestController.java]: or I'll move the logic from here in a query that modifies the db. -- I'll modify this

File: src/main/java/meeseeks/box/controller/JobController.java (hunk 1 above)
Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/JobController.java]: @decebaldecebal this is a constructor function - not a filed (everything should be indent it at the same level)!

Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/JobController.java]: @decebaldecebal this is a constructor function - not a filed (everything should be indent it at the same level)!

File: src/main/java/meeseeks/box/controller/RequestController.java (hunk 2 above)
Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/RequestController.java]: I will kill it when I'll move the logic from Controllers to Services -- I think I'll do that later today.

File: src/main/java/meeseeks/box/controller/RequestController.java (hunk 3 above)
Comment (alexandrustoica) [reply src/main/java/meeseeks/box/controller/RequestController.java]: yea, I will solve this by making Hibernate and Spring take care of Auth.Principle
